Multiple data path shard creation bug fix.

howardhuanghua opened the pull request:
# Issue
In logging/metrics scenarios, we need to store huge amount of data (e.g. TB/PB) in single cluster. Most of the history data would be cold data, in order to lower down the cost, normally customers choose multiple large HDD storages, like 4TB HDD * 12 = total 48TB. In this case, if we use like RAID0 to group 12 disks, if one of the disks down, we need to recovery whole 48TB data. So we use multiple data path on each single disk instead.

After using multiple data paths, we found an issue:
Setup 1 single node cluster with 12 data paths configured (/data1 - /data12).
Create 2000 indices, and each index has 5 shards no replica, total 10000 shards.
We could see that all indices/shards are allocated to only 5 data paths:
```
   curl "localhost:9200/_cat/shards?h=path.data" |sort |uniq -c

   2000 /data10/containers/1573522697756265109/es/data/nodes/0
   2000  /data3/containers/1573522697756265109/es/data/nodes/0
   2000  /data7/containers/1573522697756265109/es/data/nodes/0
   2000  /data8/containers/1573522697756265109/es/data/nodes/0
   2000  /data9/containers/1573522697756265109/es/data/nodes/0
```
We expect all the shards could be balanced between each path, otherwise would have performance issue.

# Analysis
I found that original logic tries to use current index's shard count per path, total shard count per path and finally left storage space three tiers to sort and select the best path:
https://github.com/elastic/elasticsearch/blob/2382c752ac339389f42d5e7fe5a5b6cb3ab261db/server/src/main/java/org/elasticsearch/index/shard/ShardPath.java#L232-L254
However, we could see that `dataPathToShardCount` has been created based on current index, **NOT** total indices:
https://github.com/elastic/elasticsearch/blob/2382c752ac339389f42d5e7fe5a5b6cb3ab261db/server/src/main/java/org/elasticsearch/index/IndexService.java#L429-L441
So shards could not be balanced between different paths.

# Solution
With this PR, I fix the shard balanced issue between multiple data paths and avoid creating/deleting extra folders during deleting single shard.
1) Introduce `numShards` in `NodeEnvironment.NodePath`, it will calculate all the loaded/created shards under this `NodePath`.
2) Introduce `indices` map in `NodeEnvironment.NodePath`, it will calculate all the loaded/created shards for each index under this `NodePath`.
3) The above statisitics number would be updated in shard or index load/create/remove operations.
4) Use above statisitics to select the best `NodePath` during creating new shard in multiple data paths case.

# Test Result
With above fix, we still create 10000 shards, we could see shards could be balanced between each data path:
```
   curl "localhost:9200/_cat/shards?h=path.data" |sort |uniq -c

   833 /data10/containers/1573522697756265109/es/data/nodes/0
   834  /data3/containers/1573522697756265109/es/data/nodes/0
   833  /data7/containers/1573522697756265109/es/data/nodes/0
   833  /data8/containers/1573522697756265109/es/data/nodes/0
   834  /data9/containers/1573522697756265109/es/data/nodes/0
   833  /data4/containers/1573522697756265109/es/data/nodes/0
   833  /data6/containers/1573522697756265109/es/data/nodes/0
   834  /data12/containers/1573522697756265109/es/data/nodes/0
   833  /data1/containers/1573522697756265109/es/data/nodes/0
   833  /data5/containers/1573522697756265109/es/data/nodes/0
   834  /data2/containers/1573522697756265109/es/data/nodes/0
   833  /data11/containers/1573522697756265109/es/data/nodes/0
```
Before fix, creating 10000 shards total cost 2910 seconds, after fix, total cost 2886 seconds. 
Shard creating performance is a little bit better, but it has the right result that we expected.

howardhuanghua committed:
Multiple data path shard creation performance optimization.

howardhuanghua committed:
make node path as final in ShardPath

howardhuanghua committed:
remove index if it exists

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Allocation)

ywelsch commented:
@howardhuanghua Why are these two changes bundled together? Also, this PR is not in reviewable form as it does not have any tests.

howardhuanghua commented:
Hi @ywelsch , I will seperate them out and add more test cases.

howardhuanghua committed:
revert select direct path

howardhuanghua committed:
add new multiple data path shard allocation test case.

howardhuanghua committed:
Merge branch 'master' into opt_multi_path

howardhuanghua commented:
Hi @ywelsch , I have removed the performance optimization part, and only keep the multiple data path shard creation bug fix. I will submit another performance optimization PR if this bug fix could be merged. Also, I have added new multiple data path creation test cases to cover three tiers of new path selection logics and removed the old one that just cover a single function, please help to check again. Thanks.

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java:
```diff
@@ -78,6 +78,7 @@
 public abstract class ESSingleNodeTestCase extends ESTestCase {
 
     private static Node NODE = null;
+    private final Path nodeDir = createTempDir();
```
I'm not sure why this change is needed.
Given that the only caller of `getNodeDir` is overriding node settings, can't you just create two temporary directories there?

howardhuanghua replied:
I thought we could centralize the node directory. I created another temporary directories instead as you suggested.

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -119,6 +123,39 @@ public NodePath(Path path) throws IOException {
             }
         }
 
+        public void addShard(ShardId shardId) {
```
do we need to make these methods concurrency-safe?

howardhuanghua replied:
I have moved caller of these methods from NodeEnv to IndexService. It's synchronized now.

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -598,16 +637,20 @@ public static void acquireFSLockForPaths(IndexSettings indexSettings, Path... sh
      * @throws ElasticsearchException if the write.lock is not acquirable
      */
-    public void deleteShardDirectoryUnderLock(ShardLock lock, IndexSettings indexSettings) throws IOException {
+    public void deleteShardDirectoryUnderLock(ShardLock lock, ShardPath shardPath, IndexSettings indexSettings) throws IOException {
```
`@Nullable ShardPath shardPath`

howardhuanghua replied:
I reverted the modification of this method.

ywelsch commented on server/src/main/java/org/elasticsearch/index/shard/ShardPath.java:
```diff
@@ -234,12 +254,10 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
                         .sorted((p1, p2) -> {
-                                int cmp = Long.compare(pathToShardCount.getOrDefault(p1, 0L),
-                                    pathToShardCount.getOrDefault(p2, 0L));
+                                int cmp = Long.compare(p1.getNumShards(indexName), p2.getNumShards(indexName));
```
Previously we used to sort by what shards are on disk, whereas now it's about allocated shards. This can interact with allocation of existing shards in subtle ways (as we allocate existing shards to the same data path where they were residing). Each of the two approaches has challenges, given the lack of balancing across data paths.

howardhuanghua replied:
Is that new shards and existing shards allocate together case I described in the above comment?

ywelsch commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -119,6 +123,39 @@ public NodePath(Path path) throws IOException {
+        }
+
+        public int getNumShards() {
+            return numShards;
```
This tracks allocated shards (i.e. for which there is an IndexService / IndexShard). I'm not sure  whether it's worth duplicating the state for IndicesService/IndexService about allocated shards here, as it introduces the risk of these two getting out of sync, and wrong book-keeping to happen here.

howardhuanghua replied:
I moved adding/removing shard operations from NodeEnv to IndexService, and add assert to make sure they always got synced in creating/deleting index or shard. 
State in NodeEnv is data path level, key is data path, whereas IndexService shards are index level.

ywelsch reviewed (changes requested):
I've left some comments. I'm not sure that the complexity with the additional state in NodeEnvironment is warranted. I also worry about how this fares in case where a node is restarted and has already a lot of (unallocated) shards on disk, which allocation of a new shard looks to be completely unaware of.

howardhuanghua committed:
Merge branch 'master' into opt_multi_path

howardhuanghua committed:
refactor based on the comment

howardhuanghua commented on server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -105,6 +105,10 @@
+        // total shard number under this NodePath
+        private int numShards = 0;
+        // shardId list per index under this NodePath
+        private final Map<String, ArrayList<ShardId>> indices = new HashMap<>();
```
I have replaced the string index name with `Index` object.

howardhuanghua commented:
Hi @ywelsch , thanks for the comment.
> I'm not sure that the complexity with the additional state in NodeEnvironment is warranted.
1. The additional `numShards` and `indices` just simple `int` and `HashMap` structures in NodeEnvironment, and only have simple add/remove operations, just like `ModelNode/ModelIndex` in `BalancedShardsAllocator`. Also, they are updated dynamically based on cluster state updating, we never need to fully rebuild it. It's different with shards in `IndexService`, shards in `IndexService` are index level, whereas shards in NodeEnv are data path level, it's similar with `RoutingTable` and `RoutingNodes`.
2. If we don't have these additional states in NodeEnv, we need to read all the shard folders from disk under this node including all the data paths in creating a new shard. This is really expensive, we have verified that if multiple data path single node has 10000 shards, HDD needs several seconds to select a new path. Current version doesn't have the performance issue due to the above described bug, it's only checked current creating index's shards num not all the shards in each data path.

> I also worry about how this fares in case where a node is restarted and has already a lot of (unallocated) shards on disk, which allocation of a new shard looks to be completely unaware of.

If we restart a data node, all shards on this node should be in `UNASSIGNED` state, they will be allocated to the original existing node and reuse original existing data path, also probably be throttled by concurrent recovery limitation. If we have new shards creation at the same allocation term, new shard could not be allocated until all existing data copy shards got allocated. 

Let me take an example to explain the details:
Let's say we have 3 nodes A/B/C, each node has 3 shards, node concurrent recovery limitation is 2, node delay allocation timeout is 1 min. We could see these steps:
1. Restart node A, 3 shards are in `UNASSIGNED` state in 1 min due to delay allocation.
2. After node A gets started, master will allocate 2 (throttled limitation) of the `UNASSIGNED` shards to node A (it has the reusable data copy) using `GatewayAllocator`, the last left `UNASSIGNED` shard we called "shardU". Meanwhile, we received a new index creation request, and if one of the new shards we called "shardN" also should be allocated to node A. Since it's throttled, **the "shardN" also cannot be allocated** and will wait for the next round allocation.
3. After 2 of above `UNASSIGNED` shards got recovered, the left "shardU" and "shardN" will be allocated together to node A. In this case, the "shardN" may be created **before** "shardU" in `IndicesClusterStateService`. This would cause "shardN" may take a not that accurate shard num state to select new data path, since "shardU" definitely goes to the path it used to be, and it hasn’t been counted yet. But it should only have a tiny impact as it only happens in the last round of throttled allocation. Also, the continuous incoming creating shards would balance the multiple data path of this node again.

Except the above case, all the shards including existing or new creating shards should be calculated by the correct shard num state in NodeEnv during selecting new path.


Another modification:
I have moved `addShard` `removeShard` caller from NodeEnvironment into IndexService `createShard` `removeShard`. It has these purposes:
1. Keep NodeEnv's shard state synced with IndexService. They remove/create shards together, and I also added extra assert to make sure all the indices and shards state are synced.
2. We only keep in-memory shard state in NodeEnv together with IndexService, if shard that only exists on disk, I think it should be removed (pending deletion) or loaded (dangling) eventually. Even it cannot be deleted or loaded, we still could allocate new shards to related paths, low/high disk watermark could control the shard relocation to avoid single data path space to be exploded. Closed index/shard also has an entry in memory.
3. We don't need to care about concurrency-safe in adding/removing shards for NodeEnv, as IndexService would take care it.

howardhuanghua commented:
Hi @ywelsch , would you please help to check again?

howardhuanghua commented:
Hi @ywelsch, shall we need to continue to discuss this bug fix or we have other approaches? Thanks.

rjernst commented:
Multiple data paths will be deprecated in 7.13 and removed in 8.0 (see https://github.com/elastic/elasticsearch/issues/71205). Given that, I believe this PR is no longer relevant, so I will close it.
